src/AfterMiddlewareCaller.php

```diff
@@ -0,0 +1,90 @@
+ *
+ * @author Felix Becker <f.becker@outlook.com>
+ */
+class AfterMiddlewareCaller
+{
+    /**
+     * @var Application
+     */
+    private $app;
+
+    /**
+     * @var ParameterResolver
+     */
+    private $parameterResolver;
+
+    /**
+     * @param Application $app the application that holds the container for injection
+     */
+    public function __construct(Application $app)
+    {
+        $this->app = $app;
+        $this->parameterResolver = new ResolverChain([
+            new TypeHintContainerResolver($app->getContainer()), // first resolve by type hints
+            new NumericArrayResolver, // fall back to Silex' parameter order
+        ]);
+    }
+
+    /**
+     * Calls the given middleware, performs injection and returns the result
+     *
+     * @param  mixed    $middleware the middleware
+     * @param  Request  $request    the request object
+     * @param  Response $response   the response object
+     * @return mixed                the return value of the middleware
+     */
+    public function call($middleware, Request $request, Response $response)
+    {
+        $middleware = $this->app['callback_resolver']->resolveCallback($middleware);
+        $middlewareReflection = CallableReflection::create($middleware);
+        $middlewareParameters = $middlewareReflection->getParameters();
+        $resolvedArguments = [];
+
+        // Check for a typehint of the request and response object at any parameter position
+        foreach ($middlewareParameters as $index => $parameter) {
+            if ($parameter->getClass()) {
+                if ($parameter->getClass()->isInstance($request)) {
+                    $resolvedArguments[$index] = $request;
+                } elseif ($parameter->getClass()->isInstance($response)) {
+                    $resolvedArguments[$index] = $response;
+                }
+            }
+        }
+
+        // resolve all other arguments through the container (this includes the application object)
+        // if no type for the first or second arguments are found, it falls back to the default Silex
+        // behaviour of passing the request, response and application
+        $arguments = $this->parameterResolver->getParameters($middlewareReflection, [$request, $response, $this->app], $resolvedArguments);
+
+        ksort($arguments);
+
+        // Check if all parameters are resolved
+        $diff = array_diff_key($middlewareParameters, $arguments);
+        if (0 < count($diff)) {
+            /** @var \ReflectionParameter $parameter */
+            $parameter = reset($diff);
+            throw new \RuntimeException(sprintf(
+                'Middleware "%s" requires that you provide a value for the "$%s" argument.',
+                $middlewareReflection->getName(),
+                $parameter->getName()
+            ));
+        }
+
+        return call_user_func_array($middleware, $arguments);
```

Comment:
I took the implementation from [ControllerResolver::getArguments](https://github.com/PHP-DI/Silex-Bridge/blob/master/src/Controller/ControllerResolver.php#L65). Does `Invoker::call()` support a custom `ResolverChain`? Because I have to fallback to a `NumericParameterResolver`